src/kernel-install/00-entry-directory.install

```diff
@@ -1,48 +0,0 @@
-#!/usr/bin/env bash
-# -*- mode: shell-script; indent-tabs-mode: nil; sh-basic-offset: 4; -*-
-# ex: ts=8 sw=4 sts=4 et filetype=sh
-# SPDX-License-Identifier: LGPL-2.1-or-later
-#
-# This file is part of systemd.
-#
-# systemd is free software; you can redistribute it and/or modify it
-# under the terms of the GNU Lesser General Public License as published by
-# the Free Software Foundation; either version 2.1 of the License, or
-# (at your option) any later version.
-#
-# systemd is distributed in the hope that it will be useful, but
-# WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
-# General Public License for more details.
-#
-# You should have received a copy of the GNU Lesser General Public License
-# along with systemd; If not, see <http://www.gnu.org/licenses/>.
-
-COMMAND="$1"
-KERNEL_VERSION="$2"
-ENTRY_DIR_ABS="$3"
-KERNEL_IMAGE="$4"
-INITRD_OPTIONS_START="5"
-
-if ! [[ $KERNEL_INSTALL_MACHINE_ID ]]; then
-    exit 0
-fi
-
-if [[ $COMMAND != add ]]; then
-    exit 0
-fi
-
-# If the boot dir exists (e.g. $ESP/<machine-id>),
-# create the entry directory ($ESP/<machine-id>/<kernel-version>).
-# This is the only function of this plugin.
-MACHINE_ID_DIR="${ENTRY_DIR_ABS%/*}"
-if ! [ -d "$MACHINE_ID_DIR" ]; then
-    exit 0
-fi
-
-if [ "$KERNEL_INSTALL_VERBOSE" -gt 0 ]; then
-    echo "+mkdir -v -p $ENTRY_DIR_ABS"
-    exec mkdir -v -p "$ENTRY_DIR_ABS"
-else
-    exec mkdir -p "$ENTRY_DIR_ABS"
-fi
```

Comment:
I don't think this can work. `/usr/lib/kernel/install.d/20-grub.install` has the following:
```sh
...
BOOT_DIR_ABS="$3"
...
# If ${BOOT_DIR_ABS} exists, some other boot loader is active.
[[ -d "${BOOT_DIR_ABS}" ]] && exit 0
```
Also `50-dracut.install` looks for this directory.

The presence of the directory was documented as the marker to tell scriptlets that the BLS is used. In hindsight, this was a bad decision, mea culpa. But we can't just drop this interface, since this will break on existing systems with scriptlets from those packages.

I now see that https://github.com/systemd/systemd/pull/16687/commits/31e57550b552e113bd3d44355b237c41e42beb58 broke things. In the PR I said "The big question is whether kernel-install will do the right thing now with grub2 installed." Unfortunately the answer is "no" :((

Instead, we can add a new interface that is more resilient and move over our stuff to the new interface. Other folks can move over too, but don't need to.

Actually, I'm not sure what this new interface should be. Should we try to perform some autodetection and set a variable like KERNEL_INSTALL_LAYOUT=... ?

Reply:
~~I think this is less grim than it might seem.~~ Okay so this sucks.

Debian, for example, provides (`apt-file search /usr/lib/kernel/install.d`):
```
systemd: /usr/lib/kernel/install.d/00-entry-directory.install
systemd: /usr/lib/kernel/install.d/50-depmod.install
systemd: /usr/lib/kernel/install.d/85-initrd.install
systemd: /usr/lib/kernel/install.d/90-loaderentry.install
```

The only interesting one here is `85-initrd.install`, but that only copies /boot/initrd.img-whatever to $BOOT_DIR_ABS/initrd, without destination existence checks, and is therefore unaffected.

According to [`yum whatprovides '/usr/lib/kernel/install.d/*'`](https://bugzilla.redhat.com/show_bug.cgi?id=1957469), Fedora ships (simplified and deduplicated):
```
akmods-0.5.6-27.fc34.noarch: Automatic kmods build and install tool
Filename: /usr/lib/kernel/install.d/95-akmodsposttrans.install

dkms-2.8.4-2.fc34.noarch: Dynamic Kernel Module Support Framework
Filename: /usr/lib/kernel/install.d/40-dkms.install

dracut-053-5.fc34.x86_64: Initramfs generator using udev
Filename: /usr/lib/kernel/install.d/50-dracut.install

dracut-config-rescue-053-5.fc34.x86_64: dracut configuration to turn on rescue image generation
Filename: /usr/lib/kernel/install.d/51-dracut-rescue.install

grub2-common-1:2.06~rc1-4.fc34.noarch: grub2 common layout
Filename: /usr/lib/kernel/install.d/20-grub.install
Filename: /usr/lib/kernel/install.d/99-grub-mkconfig.install

grubby-8.40-51.fc34.x86_64: Command line tool for updating bootloader configs
Filename: /usr/lib/kernel/install.d/10-devicetree.install
Filename: /usr/lib/kernel/install.d/95-kernel-hooks.install

kexec-tools-2.0.21-5.fc34.x86_64: The kexec/kdump userspace component
Filename: /usr/lib/kernel/install.d/60-kdump.install

systemd-udev-248-2.fc34.x86_64: Rule-based device node and kernel event manager
Filename: /usr/lib/kernel/install.d/00-entry-directory.install
Filename: /usr/lib/kernel/install.d/20-grubby.install
Filename: /usr/lib/kernel/install.d/50-depmod.install
Filename: /usr/lib/kernel/install.d/90-loaderentry.install

tuned-2.15.0-2.fc34.noarch: A dynamic adaptive system tuning daemon
Filename: /usr/lib/kernel/install.d/92-tuned.install
```

I, naturally, installed all of them. [Here](https://gist.github.com/nabijaczleweli/d98ba80f3d0a3ecb4e09e28fea46fea6)'s a gist.

This ends up with
```
95-kernel-hooks.install:BOOT_DIR_ABS="$3"
10-devicetree.install:#BOOT_DIR_ABS="$3"
95-akmodsposttrans.install:BOOT_DIR_ABS="$3"
92-tuned.install:BOOT_DIR_ABS="$3"
60-kdump.install:BOOT_DIR_ABS="$3"
50-dracut.install:BOOT_DIR_ABS="$3"
90-loaderentry.install:ENTRY_DIR_ABS="$3"
50-depmod.install:ENTRY_DIR_ABS="$3"
20-grubby.install:BOOT_DIR_ABS="$3"
00-entry-directory.install:ENTRY_DIR_ABS="$3"
20-grub.install:BOOT_DIR_ABS="$3"
```
or, if we get rid of ones from systemd:
```
51-dracut-rescue.install:BOOT_DIR_ABS="${3%/*}/0-rescue"
51-dracut-rescue.install:if [[ -d "${BOOT_DIR_ABS%/*}" ]]; then
51-dracut-rescue.install:    BOOT_ROOT=${BOOT_DIR_ABS%$BOOT_DIR}
51-dracut-rescue.install:    BOOT_DIR_ABS="/boot"
51-dracut-rescue.install:        [[ -f "$LOADER_ENTRY" ]] && [[ -f "$BOOT_DIR_ABS/$KERNEL" ]] \
51-dracut-rescue.install:            && [[ -f "$BOOT_DIR_ABS/$INITRD" ]] && exit 0
51-dracut-rescue.install:        [[ -d "$BOOT_DIR_ABS" ]] || mkdir -p "$BOOT_DIR_ABS"
51-dracut-rescue.install:        if ! cp --reflink=auto "$KERNEL_IMAGE" "$BOOT_DIR_ABS/$KERNEL"; then
51-dracut-rescue.install:            echo "Can't copy '$KERNEL_IMAGE to '$BOOT_DIR_ABS/$KERNEL'!" >&2
51-dracut-rescue.install:        if [[ ! -f "$BOOT_DIR_ABS/$INITRD" ]]; then
51-dracut-rescue.install:            dracut -f --no-hostonly -a "rescue" "$BOOT_DIR_ABS/$INITRD" "$KERNEL_VERSION"
51-dracut-rescue.install:        if [[ "${BOOT_DIR_ABS}" != "/boot" ]]; then
95-kernel-hooks.install:BOOT_DIR_ABS="$3"
95-kernel-hooks.install:# If $BOOT_DIR_ABS exists, some other boot loader is active.
95-kernel-hooks.install:[[ -d "$BOOT_DIR_ABS" ]] && exit 0
10-devicetree.install:#BOOT_DIR_ABS="$3"
95-akmodsposttrans.install:BOOT_DIR_ABS="$3"
92-tuned.install:BOOT_DIR_ABS="$3"
60-kdump.install:BOOT_DIR_ABS="$3"
60-kdump.install:if [[ -d "$BOOT_DIR_ABS" ]]; then
60-kdump.install:    BOOT_DIR_ABS="/boot"
60-kdump.install:        rm -f -- "$BOOT_DIR_ABS/$KDUMP_INITRD"
50-dracut.install:BOOT_DIR_ABS="$3"
50-dracut.install:# If KERNEL_INSTALL_MACHINE_ID is defined but empty, BOOT_DIR_ABS is a fake directory.
50-dracut.install:if [[ -d "$BOOT_DIR_ABS" ]]; then
50-dracut.install:    BOOT_DIR_ABS="/boot"
50-dracut.install:            cp --reflink=auto "$INITRD_IMAGE_PREGENERATED" "$BOOT_DIR_ABS/$INITRD" \
50-dracut.install:                && chown root:root "$BOOT_DIR_ABS/$INITRD" \
50-dracut.install:                && chmod 0600 "$BOOT_DIR_ABS/$INITRD" \
50-dracut.install:	dracut -f ${noimageifnotneeded:+--noimageifnotneeded} "$BOOT_DIR_ABS/$INITRD" "$KERNEL_VERSION"
50-dracut.install:        rm -f -- "$BOOT_DIR_ABS/$INITRD"
20-grubby.install:BOOT_DIR_ABS="$3"
20-grub.install:BOOT_DIR_ABS="$3"
20-grub.install:# If ${BOOT_DIR_ABS} exists, some other boot loader is active.
20-grub.install:[[ -d "${BOOT_DIR_ABS}" ]] && exit 0
```

This is great! Half of these use the presence of this inconsequential directory as direxion to either write There or do something completely different. `50-dracut.install` either writes under the directory we it's provided or under `/boot`! Cool! I assume this is what you mean by "BLS"?

(The `20-grub.install` approach is the simplest to correct. It's obviously wrong and can just be removed, in lieu of a sensible way of disabling it by linking it to /dev/null in /etc. Others, Very Not So.)

I'm not quite sure what to do about this. My primordial instinct is "to hell with fedora, this is self-inflicted", but, you know, I doubt most others here would agree with me :P

Hm. Maybe doing something like moving
```sh
export KERNEL_INSTALL_USE_BLS=
[ -d "$BOOT_ROOT/$MACHINE_ID" ] && KERNEL_INSTALL_USE_BLS=yes

if [ -n "$KERNEL_INSTALL_USE_BLS" ]; then
    mkdir -v -p "$ENTRY_DIR_ABS"
fi
```
into kernel-install, with the top bit being used as the official signaling to either use or not use "BLS" (not sure what that is still, so the name is shit as a consequence). And the bottom bit serving as compat, essentially?

Downstreams could then move to checking for that variable to decide to do something different if they want to, so the check in 90-loaderentry becomes `[ -n "$KERNEL_INSTALL_USE_BLS" ] || exit 0`?

a puzzle indeed

Reply:
Updated to include this at the top (previous version [here](https://github.com/nabijaczleweli/systemd/tree/bootctlpsko-v4)). Thoughts?